### PR TITLE
[10.0][CHG] being able to make the difference between an update done by the customer and update done internally in Odoo

### PR DESCRIPTION
**Pull request by Cedric-Pigeon**
refers to https://github.com/shopinvader/odoo-shopinvader/issues/590

### CODE REVIEW

File: shopinvader/models/track_external_update.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    _name = "track.external.update"
+    _description = "Track external update"
+
+    last_external_update_date = fields.Datetime()
+
+    def is_called_from_rest(self):
+        return self._context.get("shopinvader_request", False)
+
+    @api.multi
+    @api.depends("write_date")
```
Comment (sebastienbeau) [inline shopinvader/models/track_external_update.py]: I thinks this code is not used

**Review by sebastienbeau — APPROVED** [review]
LGTM code review
thanks for the improvement

File: shopinvader/models/track_external_update.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
+# -*- coding: utf-8 -*-
+# Copyright 2020 ACSONE SA/NV (<http://acsone.eu>)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+from odoo import api, fields, models
+
+
+class TrackExternalUpdate(models.AbstractModel):
+    _name = "track.external.update"
```
Comment (simahawk) [inline shopinvader/models/track_external_update.py]: I'd call it `track.external.mixin`. Update or not it's a mixin, we could add other things unrelated to updates in the future.

File: shopinvader/models/track_external_update.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    _description = "Track external update"
+
+    last_external_update_date = fields.Datetime()
+
+    def is_called_from_rest(self):
+        return self._context.get("shopinvader_request", False)
```
Comment (simahawk) [inline shopinvader/models/track_external_update.py]:
```suggestion
        return self.env.context.get("shopinvader_request", False)
```

File: shopinvader/services/service.py [hunk 4]
```diff
@@ -19,6 +19,14 @@ class BaseShopinvaderService(AbstractComponent):
     _collection = "shopinvader.backend"
     _expose_model = None
 
+    @property
+    def env(self):
+        env = self.work.env
+        context = dict(env.context)
+        context["shopinvader_request"] = True
+        env.context = context
```
Comment (simahawk) [inline shopinvader/services/service.py]: I think you should convert it back to `frozendict`.
Should be something like:
```py
env.context = frozendict(dict(env.context, shopinvader_request=True))
```
Also, I wonder if we should not set this directly in the env of the collection instead, where this env comes from.

File: shopinvader/models/track_external_update.py [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+from odoo import api, fields, models
+
+
+class TrackExternalUpdate(models.AbstractModel):
+    _name = "track.external.update"
+    _description = "Track external update"
+
+    last_external_update_date = fields.Datetime()
+
+    def is_called_from_rest(self):
```
Comment (simahawk) [inline shopinvader/models/track_external_update.py]: `is_rest_request`?

**Review by simahawk — CHANGES_REQUESTED** [review]

File: shopinvader/services/service.py (hunk 4 above)
Comment (Cedric-Pigeon) [reply shopinvader/services/service.py]: @simahawk I tried to do it for hours but did not succeed without recreating a new environment.

File: shopinvader/models/track_external_update.py [hunk 6]
```diff
@@ -0,0 +1,26 @@
+# -*- coding: utf-8 -*-
```
Comment (simahawk) [inline shopinvader/models/track_external_update.py]: only thing left: rename the file ;)

**Review by simahawk — APPROVED** [review]

**Review by acsonefho — APPROVED** [review]
LGTM code review! Seems great!

**Pull request merged**